sources/api/migration/migration-helpers/src/common_migrations.rs

```diff
@@ -1260,6 +1260,114 @@ mod test_add_metadata {
 
 // =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
+/// We use this migration when we remove metadata
+pub struct RemoveMetadataMigration(pub &'static [SettingMetadata]);
+
+impl Migration for RemoveMetadataMigration {
+    fn forward(&mut self, mut input: MigrationData) -> Result<MigrationData> {
+        for setting_metadata in self.0 {
+            if let Some(found_metadata) = input.metadata.get_mut(setting_metadata.setting) {
+                for metadata in setting_metadata.metadata {
+                    if let Some(metadata_value) = found_metadata.remove(*metadata) {
+                        println!(
+                            "Removed {}, which was set to '{}'",
+                            metadata, metadata_value
+                        );
+                    } else {
+                        println!(
+                            "Found no metadata '{}' to remove on setting '{}'",
+                            metadata, setting_metadata.setting
+                        );
+                    }
+                }
+            } else {
+                println!(
+                    "Found no metadata for '{}' setting",
+                    setting_metadata.setting
+                );
+            }
+        }
+        Ok(input)
+    }
+
+    fn backward(&mut self, input: MigrationData) -> Result<MigrationData> {
+        println!(
+            "RemoveMetadataMigration({:?}) has no work to do on downgrade.",
```

Comment:
Double-checking this is right, @bcressey. We don't need to do anything on downgrade because the defaults will cause the metadata to be put back into place on downgrade, right?

Reply:
> Double-checking this is right, @bcressey. We don't need to do anything on downgrade because the defaults will cause the metadata to be put back into place on downgrade, right?

Yes. `migrator.service` runs before `storewolf.service` and (simplifying) is responsible for fixing things up so the datastore that `storewolf` deserializes matches what the model expects.

`storewolf` will add any new defaults to an existing data store (since 036be7876b7e92f8f50063b5faded8e799912ab5) but does not touch existing data.

One conclusion I've drawn from this - frankly downright weird! - migration is that there's probably something much smarter we could do for the datastore files that cannot be set via the API, such as "remove all existing metadata so that `storewolf` puts it back".

But that would fall outside the scope of what we can cover in any one set of up/down migrations; we'd need the OS itself to be smart enough to do that, which means doing the repo pivot trick to move to the new world where we can treat it as a given.

If we did that, we could expand the scope to include configuration-files and services too. Might as well throw in unrecognized prefixes and individual files while we're at it, and the ability to set a field to `{{ default }}` to force it to be re-evaluated to its current default value. Potentially we could eliminate almost all of what we use migrations for even in the current settings system.